tests/trainer/test_trainer.py

```diff
@@ -177,8 +188,12 @@ def _optimizer_step(epoch, batch_idx, optimizer, optimizer_idx,
                     # use this opportunity to test once
                     assert trainer.accumulate_grad_batches == expected[2]
 
-                assert batch_idx == model.prev_called_batch_idx
-                model.prev_called_batch_idx += expected[2]
+                if trainer.num_training_batches == batch_idx + 1:
+                    # prev_called_batch_idx - schedule + modulus remainder
+                    assert batch_idx == (model.prev_called_batch_idx - expected[2] + (batch_idx + 1) % expected[2])
+                else:
+                    assert batch_idx == model.prev_called_batch_idx
+                    model.prev_called_batch_idx += expected[2]
```

Comment:
@awaelchli now it's shorter, but still has complexity

```py
@pytest.mark.parametrize(
    ['schedule', 'expected'],
    [
        pytest.param({1: 2, 3: 4}, [1, 2, 2, 4]),  # make 4 items for 4 epochs
        pytest.param(3, [3, 3, 3, 3]),
        pytest.param(4, [4, 4, 4, 4])
    ]
)
def test_gradient_accumulation_scheduling(tmpdir, schedule, expected):
  model = EvalModelTemplate()
  trainer = Trainer(
    accumulate_grad_batches=schedule,
    limit_train_batches=0.7,  # not to be divisible by accumulate_grad_batches on purpose
    limit_val_batches=0.8,
    max_epochs=4,
    default_root_dir=tmpdir
  )

  # test optimizer call freq matches scheduler
  def _optimizer_step(epoch, batch_idx, optimizer, optimizer_idx,
                      second_order_closure=None, on_tpu=False,
                      using_native_amp=False, using_lbfgs=False):

    if epoch in range(trainer.max_epochs):
      # epoch - [0, 1, 2, 3]
      if batch_idx == expected[epoch] - 1:
        model.prev_called_batch_idx = expected[epoch] - 1
        assert trainer.accumulate_grad_batches == expected[epoch]

        # separate check for 1 accumulate step at the end of batch
        if expected[epoch] == 1 and trainer.num_training_batches == batch_idx + 1:
          assert batch_idx == model.prev_called_batch_idx
        elif trainer.num_training_batches == batch_idx + 1:

          # since model.prev_called_batch_idx has accmulation factor added
          # in the below `else` block, need to subtract it
          # and add the last batch (which is modulus remainder)

          assert batch_idx == (model.prev_called_batch_idx - expected[epoch] \
                                         + (batch_idx + 1) % expected[epoch])
        else:
          assert batch_idx == model.prev_called_batch_idx
          model.prev_called_batch_idx += expected[epoch]
        
        optimizer.step()

        # clear gradients
        optimizer.zero_grad()

    # for the test
    model.optimizer_step = _optimizer_step
    model.prev_called_batch_idx = 0

    trainer.fit(model)
```

Reply:
I think to be safe I would keep it as is. It would be better to make these modifications in a new PR. I think with bugfixes it is important to change tests as little as possible and only add new tests, so with the current state I think we satisfy that :)

Reply:
Alright.